### PR TITLE
feat(workspace): option to gen title using full hierarchy

### PR DESCRIPTION
**Pull request by jonathanyeung**
## feat(workspace): option to gen title using full hierarchy

Gives a new workspace config setting to allow users to utilize the full hierarchy to generate the note title upon note creation. For example,  a new note named `one.two.three` will generate a note title of `One Two Three` as opposed to just `Three`. 

Docs PR: https://github.com/dendronhq/dendron-site/pull/453

---

# Dendron Extended PR Checklist

- NOTE: the links don't work. you'll need to go into the wiki and use lookup to find the note until we fix some issues in the markdown export

### Basics

- [x] code should follow [Code Conventions](dev.process.code)
- [ ] circular dependency check: make sure your code is not introducing new circular dependencies in plugin-core.  See [Avoiding Circular Dependencies](dev.process.code.best-practices).
- [x] sticking to existing conventions instead of creating new ones
  - eg: [if configuration for utilities are already in one module or package, add future utilities there as well](https://github.com/dendronhq/dendron/pull/1960#discussion_r786228021)

### Extended
- General
  - [x] check whether code be simplified
  - [x] check if similar function already exist in the codebase. if so, can it be re-used?
  - [x] check if this change adversely impact performance
- Operations
  - [x] when shipping this change, will it just work or will it introduce additional operational overhead due to complicated interface or known bugs?
- Architecture
  - [x] check if code is introducing changes on a foundational class or interface. if so, call for design review if needed
    - eg: [making changes to DNode](https://github.com/dendronhq/dendron/pull/2158#pullrequestreview-854689586)

### Basics
- [ ] if you are adding analytics related changes, make sure the [Telemetry](https://wiki.dendron.so/notes/84df871b-9442-42fd-b4c3-0024e35b5f3c.html) docs are updated

### Extended
- [ ] can we track the performance of this change to know if it is _successful_?
  - eg: [see usage for export pod](https://github.com/dendronhq/dendron/pull/2190#pullrequestreview-855715612)

### Basics

- [x] [Write Tests](dev.process.qa.test) 
- [ ] [Confirm existing tests pass](dev.process.qa.test)
- [x] [Confirm manual testing](dev.process.qa.test) 
- [x] Common cases tested
- [x] 1-2 Edge cases tested
- [ ] If your tests changes an existing snapshot, snapshots have been [updated](dev.process.qa.test)

### Extended
- [ ] If you are adding a new language feature (graphically visible in VS Code/preview/publishing), an example is included in the [test workspace](dev.ref.test-workspace)
- CSS
  - [ ] display is correct for following dimensions
    - [ ] sm: screen ≥ 576px, eg. iphonex, (375x812)
    - [ ] lg: screen ≥ 992px
    - [ ] xxl: screen ≥ 1600px eg. mac (1600x900)
  - [ ] display is correct for following browsers (across the various dimensions)
    - [ ] safari
    - [ ] firefox
    - [ ] chrome

## Docs
- [x] if your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR
https://github.com/dendronhq/dendron-site/pull/453
- [ ] does this change introduce a new or better way of doing things that others need to be aware of? if so, an async should be created and a process added in [Development](dev) or [Packages](pkg)

### Extended
- [ ]  is this a developer BREAKING change? if another person cloning from this branch will need to adjust their dependencies or mental model of the architecture, then it is. if this is the case, make sure this is communicated according to [Close Loop](dev.process.close-loop)
  - eg: [breaking dev change due to new dependency](https://github.com/dendronhq/dendron/pull/2188#pullrequestreview-855696330)

### CODE REVIEW

**Comment by jonathanyeung** [comment]
@kevinslin - let me know what you think about this implementation for supporting full hierarchy in the note title. 

I opted not to go with a note-traits implementation for several reasons:
1. To me, this kind of functionality isn't the intention of note traits.  Note traits are meant to apply behavior to a type of notes, whereas in this case this is workspace-wide behavior and unrelated to the type of note.  Note-trait specific behavior would/could override this workspace-wide behavior
2. Some more work needs to be done w/ note-traits first to support this scenario properly.
3. This is a much simpler implementation. 


Let me know what you think - if this is fine, I can go ahead and add tests, docs, etc.

**Comment by kevinslin** [comment]
tldr: I'm fine with this being an 80/20 fix for now but absolutely believe this behavior is better suited for note traits

To respond to your points
> 1. To me, this kind of functionality isn't the intention of note traits. Note traits are meant to apply behavior to a type of notes, whereas in this case this is workspace-wide behavior and unrelated to the type of note. Note-trait specific behavior would/could override this workspace-wide behavior
I disagree. I think note traits are to add behavior to notes, the workspace config is to set defaults. It actually makes sense for the workspace to have default traits that are auto applied and is a natural extension of traits. 

2 & 3 I agree with. 

Most of my thoughts were already captured in the task note ->  [[Change Default Note Title|dendron://private/task.2022.02.01.change-default-note-title]]

File: packages/plugin-core/src/commands/NoteLookupCommand.ts [hunk 1]
```diff
@@ -422,6 +422,10 @@ export class NoteLookupCommand
                 item.traits = [journalTrait];
               }
             }
+          } else if (
```
[the remark this answers is not in the document]
Comment (jonathanyeung) [reply packages/plugin-core/src/commands/NoteLookupCommand.ts:425]: good point, let me make an adjustment for this.

[the remark this answers is not in the document]
Comment (kevinslin) [reply packages/plugin-core/src/commands/NoteLookupCommand.ts:425]: the problem with putting the logic in plugin-core is that it won't apply to notes created via goto note or the CLI. for settings that affect note state, can we add the logic in the engine? eg. engine.writeNote

**Review by kevinslin — CHANGES_REQUESTED** [review]

File: packages/plugin-core/src/commands/NoteLookupCommand.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jonathanyeung) [reply packages/plugin-core/src/commands/NoteLookupCommand.ts:425]: It's not easy to do that right now - there isn't a way to pass in info to `DendronEngineClient` regarding journal/scratch/task flags, note-traits, or anything like that right now.  We'll have to do some re-working of how these modifiers work in order to properly support them in both CLI and plugin.

I'm inclined to just support in plugin right now, or else add some replicated logic in CLI side, which is on parity with how journal notes also work right now.

[the remark this answers is not in the document]
Comment (jonathanyeung) [reply packages/plugin-core/src/commands/NoteLookupCommand.ts:425]: @SeriousBug  - actually on second thought, I think it makes more sense to not apply this to journal notes.  Right now, journal notes will only use a date portion for the title, so I don't think it makes sense to alter that behavior.  I.e. a journal note with fname 'foo.bar.2020.01.01' produces a note with title '2020-01-01', so it doesn't make sense to alter this title to anything like 'Foo Bar 2020-01-01' since the non-numerical portion was already getting dropped.  I'll update the docs.
